api/rosetta/payloads.go

```diff
@@ -0,0 +1,127 @@
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#response-5
+type PayloadsResponse struct {
+	Transaction string                  `json:"unsigned_transaction"`
+	Payloads    []object.SigningPayload `json:"payloads"`
+}
+
+// Payloads implements the /construction/payloads endpoint of the Rosetta Construction API.
+// It receives an array of operations and all other relevant information required to construct
+// an unsigned transaction. Operations must deterministically describe the intent of the
+// transaction. Besides the unsigned transaction text, this endpoint also returns the list
+// of payloads that should be signed.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructionpayloads
+func (c *Construction) Payloads(ctx echo.Context) error {
+
+	var req PayloadsRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	// Metadata object is the response from our metadata endpoint. Thus, the object
+	// should be okay, but let's validate it anyway.
+	blockID := req.Metadata.ReferenceBlockID
+	if blockID.Index == nil && blockID.Hash == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockEmpty))
+	}
+	if blockID.Hash != "" && len(blockID.Hash) != hexIDSize {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockLength,
+			withDetail("have_length", len(blockID.Hash)),
+			withDetail("want_length", hexIDSize),
+		))
+	}
+
+	intent, err := c.parser.DeriveIntent(req.Operations)
+	var iaErr failure.InvalidAccount
+	if errors.As(err, &iaErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidAccount(iaErr))
+	}
+	var icErr failure.InvalidCurrency
+	if errors.As(err, &icErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidCurrency(icErr))
+	}
+	var ucErr failure.UnknownCurrency
+	if errors.As(err, &ucErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, unknownCurrency(ucErr))
+	}
+	var opErr failure.InvalidOperations
+	if errors.As(err, &opErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidFormat(txInvalidOps))
+	}
+	var inErr failure.InvalidIntent
+	if errors.As(err, &inErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidIntent(inErr))
+	}
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, internal(intentDetermination, err))
+	}
+
+	tx, err := c.parser.CompileTransaction(intent, req.Metadata)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, internal(txConstruction, err))
+	}
+
+	enc, err := json.Marshal(tx)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, internal(txConstruction, err))
```

Comment:
The message here should not be `txConstruction`, should it? It's an encoding error. For me, this kind of error that should never happen can just be a generic internal error, with the error explaining the details.